File: pixiv_object/token.py
from dataclasses import dataclass

from pixiv_object.pixiv_object import PixivObject


# region other classes
@dataclass
class ProfileImageUrls:
    px_16x16: str
    px_50x50: str
    px_170x170: str


@dataclass
class User:
    profile_image_urls: ProfileImageUrls
    id: int
    name: str
    account: str
    mail_address: str
    is_premium: bool
    x_restrict: int
    is_mail_authorized: bool

# endregion


@dataclass
class Token(PixivObject):
    # region field

    access_token: str
    expires_in: int
    token_type: str
    scope: str
    refresh_token: str
    user: User
    # endregion

    @staticmethod
    def object_hook(d: dict):
        # reaches highest level?
        if 'access_token' in d:
            # convert dicts to their types
            d['user']['profile_image_urls'] = ProfileImageUrls(**d['user']['profile_image_urls'])
            d['user'] = User(**d['user'])
        return d

File: src/types/token.ts
import { PixivObject } from './pixiv-object';

type User = {
	profile_image_urls: ProfileImageUrls;
	id: number;
	name: string;
	account: string;
	mail_address: string;
	is_premium: boolean;
	x_restrict: number;
	is_mail_authorized: boolean;
}

type ProfileImageUrls = {
	px_16x16: string;
	px_50x50: string;
	pi_170x170: string;
}

export type Token = PixivObject & {
	access_token: string;
	expires_in: number;
	token_type: string;
	scope: string;
	refresh_token: string;
	user: User;
}
